File: articles/design-thinking-and-systems-thinking/go/system_design_value_engine.go
package main

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
)

type Intervention struct {
	Name              string
	HumanValue        float64
	SystemLeverage    float64
	Feasibility       float64
	EquitySensitivity float64
	Durability        float64
	Risk              float64
	SystemDesignValue float64
}

func parseFloat(value string) float64 {
	parsed, err := strconv.ParseFloat(value, 64)
	if err != nil {
		log.Fatalf("unable to parse numeric value %q: %v", value, err)
	}
	return parsed
}

func systemDesignValue(x Intervention) float64 {
	return 0.24*x.HumanValue +
		0.26*x.SystemLeverage +
		0.18*x.Feasibility +
		0.14*x.EquitySensitivity +
		0.12*x.Durability -
		0.06*x.Risk
}

func loadPortfolio(path string) []Intervention {
	file, err := os.Open(path)
	if err != nil {
		log.Fatalf("unable to open input file: %v", err)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	rows, err := reader.ReadAll()
	if err != nil {
		log.Fatalf("unable to read CSV: %v", err)
	}

	var items []Intervention

	for i, row := range rows {
		if i == 0 {
			continue
		}
		if len(row) < 9 {
			log.Fatalf("malformed row: %v", row)
		}

		x := Intervention{
			Name:              row[0],
			HumanValue:        parseFloat(row[3]),
			SystemLeverage:    parseFloat(row[4]),
			Feasibility:       parseFloat(row[5]),
			EquitySensitivity: parseFloat(row[6]),
			Durability:        parseFloat(row[7]),
			Risk:              parseFloat(row[8]),
		}
		x.SystemDesignValue = systemDesignValue(x)
		items = append(items, x)
	}

	return items
}

func main() {
	path := "../data/raw/system_intervention_portfolio_raw.csv"
	if len(os.Args) > 1 {
		path = os.Args[1]
	}

	items := loadPortfolio(path)

	sort.Slice(items, func(i, j int) bool {
		return items[i].SystemDesignValue > items[j].SystemDesignValue
	})

	fmt.Println("rank,intervention,system_design_value,risk")
	for i, x := range items {
		fmt.Printf("%d,%s,%.6f,%.6f\n", i+1, x.Name, x.SystemDesignValue, x.Risk)
	}
}

File: articles/design-thinking-and-systems-thinking/c/system_design_value_engine.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_ITEMS 128
#define MAX_LINE 512
#define MAX_NAME 180

typedef struct {
    char name[MAX_NAME];
    double human_value;
    double system_leverage;
    double feasibility;
    double equity_sensitivity;
    double durability;
    double risk;
    double system_design_value;
} Intervention;

double compute_system_design_value(const Intervention *x) {
    return 0.24 * x->human_value +
           0.26 * x->system_leverage +
           0.18 * x->feasibility +
           0.14 * x->equity_sensitivity +
           0.12 * x->durability -
           0.06 * x->risk;
}

int compare_desc(const void *a, const void *b) {
    const Intervention *ia = (const Intervention *)a;
    const Intervention *ib = (const Intervention *)b;

    if (ia->system_design_value < ib->system_design_value) return 1;
    if (ia->system_design_value > ib->system_design_value) return -1;
    return 0;
}

int main(int argc, char **argv) {
    const char *path = argc > 1 ? argv[1] : "../data/raw/system_intervention_portfolio_raw.csv";
    FILE *file = fopen(path, "r");

    if (!file) {
        fprintf(stderr, "ERROR: Unable to open input file: %s\n", path);
        return 1;
    }

    Intervention interventions[MAX_ITEMS];
    char line[MAX_LINE];
    int count = 0;

    fgets(line, sizeof(line), file); /* header */

    while (fgets(line, sizeof(line), file) && count < MAX_ITEMS) {
        char *token = strtok(line, ",");
        if (!token) continue;

        strncpy(interventions[count].name, token, MAX_NAME - 1);
        interventions[count].name[MAX_NAME - 1] = '\0';

        strtok(NULL, ","); /* intervention_type */
        strtok(NULL, ","); /* leverage_level */

        token = strtok(NULL, ",");
        interventions[count].human_value = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        interventions[count].system_leverage = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        interventions[count].feasibility = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        interventions[count].equity_sensitivity = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        interventions[count].durability = token ? atof(token) : 0.0;

        token = strtok(NULL, ",");
        interventions[count].risk = token ? atof(token) : 0.0;

        interventions[count].system_design_value = compute_system_design_value(&interventions[count]);

        count++;
    }

    fclose(file);

    qsort(interventions, count, sizeof(Intervention), compare_desc);

    printf("rank,intervention,system_design_value,risk\n");
    for (int i = 0; i < count; i++) {
        printf("%d,%s,%.6f,%.6f\n", i + 1, interventions[i].name, interventions[i].system_design_value, interventions[i].risk);
    }

    return 0;
}
